client/src/modules/list/model/reducer.js

```diff
@@ -9,79 +9,64 @@ import {
 import { CohortActionTypes } from 'modules/cohort/model/actionTypes';
 import items from 'modules/list/components/Items/model/reducer';
 
-const membersReducer = (state = [], action) => {
+const membersReducer = (state = {}, action) => {
   switch (action.type) {
     case ListActionTypes.ADD_VIEWER_SUCCESS: {
       const {
         payload: { data }
       } = action;
 
-      return [...state, data];
+      return { [data._id]: data, ...state };
     }
-    case ListActionTypes.REMOVE_MEMBER_SUCCESS: {
+    case ListActionTypes.REMOVE_MEMBER_SUCCESS:
+    case ListActionTypes.LEAVE_SUCCESS: {
       const {
         payload: { userId }
       } = action;
 
-      return state.filter(member => member._id !== userId);
+      const { [userId]: deleted, ...rest } = state;
+
+      return rest;
     }
     case ListActionTypes.ADD_OWNER_ROLE_SUCCESS: {
       const {
         payload: { userId }
       } = action;
 
-      return state.map(member =>
-        member._id === userId
-          ? {
-              ...member,
-              isOwner: true,
-              isMember: true
-            }
-          : member
-      );
+      return {
+        ...state,
+        [userId]: { ...state[userId], isOwner: true, isMember: true }
+      };
     }
     case ListActionTypes.REMOVE_OWNER_ROLE_SUCCESS: {
       const {
         payload: { userId }
       } = action;
 
-      return state.map(member =>
-        member._id === userId
-          ? {
-              ...member,
-              isOwner: false
-            }
-          : member
-      );
+      return {
+        ...state,
+        [userId]: { ...state[userId], isOwner: false }
+      };
     }
     case ListActionTypes.ADD_MEMBER_ROLE_SUCCESS: {
       const {
         payload: { userId }
       } = action;
 
-      return state.map(member =>
-        member._id === userId
-          ? {
-              ...member,
-              isMember: true
-            }
-          : member
-      );
+      return {
+        ...state,
+        [userId]: { ...state[userId], isMember: true }
+      };
     }
     case ListActionTypes.REMOVE_MEMBER_ROLE_SUCCESS: {
       const {
         payload: { userId }
       } = action;
 
-      return state.map(member =>
-        member._id === userId
-          ? {
-              ...member,
-              isMember: false,
-              isOwner: false
-            }
-          : member
-      );
+      return {
+        ...state,
+        [userId]: { ...state[userId], isOwner: false, isMember: false }
```

Comment:
if member role is removed then iwner role is also removed? What i n case if we are the only owner and we remove member role?

Reply:
We have to be a member if we want to be an owner. We can't have a higher level role without having a lower level one. We can't remove a member role when we are the only owner.